nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchHttp.java

```diff
@@ -0,0 +1,330 @@
+                    sb.append(docType);
+                    sb.append("\", \"_id\": \"");
+                    sb.append(id);
+                    sb.append("\" }\n");
+                }
+            }
+            if (!flowFilesToTransfer.isEmpty()) {
+                RequestBody requestBody = RequestBody.create(MediaType.parse("application/json"), sb.toString());
+                final Response getResponse = sendRequestToElasticsearch(okHttpClient, url, "PUT", requestBody);
+                final int statusCode = getResponse.code();
+
+                if (isSuccess(statusCode)) {
+                    ResponseBody responseBody = getResponse.body();
+                    final byte[] bodyBytes = responseBody.bytes();
+                    JsonNode responseJson = parseJsonResponse(new ByteArrayInputStream(bodyBytes));
+                    boolean errors = responseJson.get("errors").asBoolean(false);
+                    if (errors) {
+                        ArrayNode itemNodeArray = (ArrayNode) responseJson.get("items");
+                        if (itemNodeArray.size() > 0) {
+                            // All items are returned whether they succeeded or failed, so iterate through the item array
+                            // at the same time as the flow file list, moving each to success or failure accordingly
+                            for (int i = 0; i < itemNodeArray.size(); i++) {
+                                JsonNode itemNode = itemNodeArray.get(i);
+                                FlowFile flowFile = flowFilesToTransfer.remove(i);
+                                int status = itemNode.findPath("status").asInt();
+                                if (!isSuccess(status)) {
+                                    String reason = itemNode.findPath("//error/reason").asText();
+                                    logger.error("Failed to insert {} into Elasticsearch due to {}, transferring to failure",
+                                            new Object[]{flowFile, reason});
+                                    session.transfer(flowFile, REL_FAILURE);
+
+                                } else {
+                                    session.transfer(flowFile, REL_SUCCESS);
+                                    // Record provenance event
+                                    session.getProvenanceReporter().send(flowFile, url.toString());
+                                }
+                            }
+                        }
+                    }
+                    // Transfer any remaining flowfiles to success
+                    session.transfer(flowFilesToTransfer, REL_SUCCESS);
+                } else {
+                    // Something went wrong during the bulk update, throw a ProcessException to indicate rollback
+                    throw new ProcessException("Received error code " + statusCode + " from Elasticsearch API");
+                }
+            }
+        } catch (IOException ioe) {
```

Comment:
This is catch statement has huge implications on whether flowfiles are inserted twice. Any way to narrow it's focus?